src/pg_copy.c

```diff
@@ -0,0 +1,696 @@
+		{
+			ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+			ShardId shardId = shardInterval->id;
+			LockShardData(shardId, ShareLock);
+			LockShardDistributionMetadata(shardId, ShareLock);
+		}
+
+		while (true)
+		{
+			ShardInterval *shardInterval = NULL;
+			ShardId shardId = 0;
+			int errorCount = 0;
+			bool found = false;
+			MemoryContext oldContext = MemoryContextSwitchTo(tupleContext);
+			nextRowFound = NextCopyFrom(copyState, NULL, columnValues, columnNulls, NULL);
+			MemoryContextSwitchTo(oldContext);
+
+			if (!nextRowFound)
+			{
+				MemoryContextReset(tupleContext);
+				break;
+			}
+
+			/* write the row to the shard */
+			if (columnNulls[partitionColumn->varattno])
+			{
+				ereport(ERROR, (errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
+								errmsg("cannot copy row with NULL value "
+									   "in partition column")));
+			}
+			partitionColumnValue = columnValues[partitionColumn->varattno];
+			hashedValue = DatumGetInt32(FunctionCall1(hashFunction,
+													  partitionColumnValue));
+			shardHashCode =
+				(int) ((uint32) (hashedValue - INT32_MIN) / hashTokenIncrement);
+			prunedList = shardListCache[shardHashCode];
+
+			if (prunedList == NULL)
+			{
+				rightConst->constvalue = partitionColumnValue;
+				prunedList = PruneShardList(tableId, whereClauseList, shardIntervalList);
+				shardListCache[shardHashCode] = prunedList;
+			}
+
+			shardInterval = (ShardInterval *) linitial(prunedList);
+			shardId = shardInterval->id;
+
+			shardConnections =
+				(ShardConnections *) hash_search(shardToConn, &shardInterval->id,
+												 HASH_ENTER,
+												 &found);
+			if (!found)
+			{
+				InitializeShardConnections(copyStatement, shardConnections, shardId,
+										   transactionManager);
+			}
+			lineBuf = CopyGetLineBuf(copyState);
+			lineBuf->data[lineBuf->len++] = '\n';
+
+			/* There was already new line in the buffer, but it was truncated:
+			 * no need to check available space */
+
+			/* Replicate row to all shard placements */
+			for (i = 0; i < shardConnections->replicaCount; i++)
+			{
+				if (PQputCopyData(shardConnections->placements[i].conn, lineBuf->data,
+								  lineBuf->len) <= 0)
+				{
+					ereport(WARNING, (errcode(ERRCODE_IO_ERROR),
+									  errmsg("Copy failed for placement %ld for %ld",
+											 (long) shardConnections->placements[i].id,
+											 (long) shardId)));
+					errorCount += 1;
+					PQfinish(shardConnections->placements[i].conn);
+					shardConnections->placements[i].conn = NULL;
+				}
+			}
+			MemoryContextReset(tupleContext);
+			shardConnections->replicaCount -= errorCount;
+			if (shardConnections->replicaCount == 0) /* if all placements failed, error out */
+			{
+				ereport(ERROR, (errcode(ERRCODE_IO_ERROR),
+								errmsg("Could not copy to any active placements for shard %ld",
+									   (long) shardId)));
+			}
+			else if (errorCount != 0) /* otherwise, mark failed placements as inactive: they're stale */
```

Comment:
Can't we roll everything back by cancelling all the COPY commands? I'm not entirely clear on the error semantics of the libpq copy functions. I think the COPY failing in-flight is a different case then failing to connect at the start. At this point we've already filtered out dead worker nodes. If something fails, it's either due to an exceptional circumstance or due to a functional failure like a constraint violation and we definitely shouldn't mark placements inactive in that case. In either case, the old behaviour of rolling back without changing placement status makes sense to me.